docs/proposals/advanced-nginx-extensions.md

```diff
@@ -26,3 +74,915 @@ NGINX use cases. To allow them to implement those use cases, we need to bring a
+type SnippetsTemplateConditionType string
+
+// SnippetsTemplateConditionReason is a reason for a SnippetsTemplate condition type.
+type SnippetsTemplateConditionReason string
+
+const (
+	// SnippetsTemplateConditionTypeAccepted indicates that the SnippetsTemplate is accepted.
+	//
+	// Possible reasons for this condition to be True:
+	//
+	// * Accepted
+	//
+	// Possible reasons for this condition to be False:
+	//
+	// * Invalid
+	// * CRDInvalid
+	// * CRDNotFound
+	// * TargetNotFound
+	SnippetsTemplateConditionTypeAccepted SnippetsTemplateConditionType = "Accepted"
+
+	// SnippetsTemplateConditionReasonAccepted is used with the Accepted condition type when
+	// the condition is true.
+	SnippetsTemplateConditionReasonAccepted SnippetsTemplateConditionReason = "Accepted"
+
+	// SnippetsTemplateConditionReasonInvalid is used with the Accepted condition type when
+	// SnippetsTemplate is invalid.
+	SnippetsTemplateConditionReasonInvalid SnippetsTemplateConditionReason = "Invalid"
+
+	// SnippetsTemplateConditionReasonCRDInvalid is used with the Accepted condition type when
+	// the referenced CRD is invalid.
+	SnippetsTemplateConditionReasonCRDInvalid SnippetsTemplateConditionReason = "CRDInvalid"
+
+	// SnippetsTemplateConditionReasonCRDNotFound is used with the Accepted condition type when
+	// the referenced CRD is not found.
+	SnippetsTemplateConditionReasonCRDNotFound SnippetsTemplateConditionReason = "CRDNotFound"
+
+	// SnippetsTemplateConditionReasonTargetNotFound is used with the Accepted condition type when
+	// the target is not found.
+	SnippetsTemplateConditionReasonTargetNotFound SnippetsTemplateConditionReason = "TargetNotFound"
+)
+```
+
+#### Templates
+
+A template must be a [go template](https://pkg.go.dev/text/template).
+
+NGF gives a template access to the data and metadata about a Custom resource (corresponds
+to `SnippetsTemplate.spec.valuesCRD`) via `TemplateContext` type:
+
+```go
+// TemplateContext provides context for NGINX configuration templates.
+type TemplateContext struct {
+	// Metadata holds the metadata for NGINX configuration templates.
+	Metadata TemplateMetadata
+	// Data holds the data for NGINX configuration templates.
+	Data TemplateData
+}
+
+// TemplateData holds the data for NGINX configuration templates.
+type TemplateData struct {
+	// Spec is the spec of the Policy/Filter of the Kind that corresponds to SnippetsTemplate.spec.valuesCRD.
+	Spec map[string]interface{}
+}
+
+// TemplateMetadata holds the metadata for NGINX configuration templates.
+type TemplateMetadata struct {
+	// TargetMeta is the metadata of the target resource like HTTPRoute.
+	TargetMeta metav1.ObjectMeta
+	// SourceMeta is the metadata of the source resource (Policy/Filter) of the Kind that corresponds to
+	// SnippetsTemplate.spec.valuesCRD.
+	SourceMeta metav1.ObjectMeta
+}
+```
+
+If requested by users, NGF can provide additional context. For example, for a template for the `location` context,
```

Comment:
How would this be requested? As a feature request?

Reply:
> As a feature request?

yep